Support cross-compiling with MinGW

nthirtyone opened the pull request:
Fixes: #71 

Toolchains and everything around them is primarily there to support CI. If you don't want that, changes in _src/system/displays/displays_windows.cpp_ and _CMakeLists.txt:148,171ff_ are enough to support MinGW.

nabijaczleweli commented on src/system/displays/displays_windows.cpp:
```diff
@@ -14,39 +14,42 @@
-		std::vector<iware::system::display_t> ret;
-	} bundle{GetDC(nullptr), {}};
-	iware::detail::quickscope_wrapper desktop_dc_deleter{[&]() { ReleaseDC(nullptr, bundle.desktop_dc); }};
+struct bundle {
```
as i understand the error, is this just an attribute(stdcall) mismatch? would passing `(MONITORENUMPROC)([](auto, ...)` as an argument instead work?

nthirtyone replied:
No, although I do not know the solid reason for it and I did not dig into it. Both c-style and reinterpret cast fail with "invalid cast":
```
/home/aki/src/infoware/src/system/displays/displays_windows.cpp: In function ‘std::vector<iware::system::display_t> iware::system::displays()’:
/home/aki/src/infoware/src/system/displays/displays_windows.cpp:26:13: error: invalid cast from type ‘iware::system::displays()::<lambda(auto:1, auto:2, auto:3, auto:4)>’ to type ‘MONITORENUMPROC’ {aka ‘int (__attribute__((stdcall)) *)(HMONITOR__*, HDC__*, tagRECT*, long int)’}
   26 |             reinterpret_cast<MONITORENUMPROC>([](auto, auto hdc, auto rect, auto userdata) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   27 |                     auto& bundle = *reinterpret_cast<struct bundle*>(userdata);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   28 |
      |
   29 |                     const unsigned int desktop_dpi = GetDeviceCaps(bundle.desktop_dc, LOGPIXELSX);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   30 |                     // https://blogs.msdn.microsoft.com/oldnewthing/20101013-00/?p=12543
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   31 |                     const unsigned int desktop_bpp    = GetDeviceCaps(bundle.desktop_dc, BITSPIXEL) * GetDeviceCaps(bundle.desktop_dc, PLANES);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   32 |                     const double desktop_refresh_rate = GetDeviceCaps(bundle.desktop_dc, VREFRESH);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   33 |
      |
   34 |
      |
   35 |                     // Sometimes returns 0 – fall back to the desktop's globals if so.
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   36 |                     const unsigned int monitor_dpi    = GetDeviceCaps(hdc, LOGPIXELSX);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   37 |                     const unsigned int monitor_bpp    = GetDeviceCaps(hdc, BITSPIXEL) * GetDeviceCaps(hdc, PLANES);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   38 |                     const double monitor_refresh_rate = GetDeviceCaps(hdc, VREFRESH);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   39 |
      |
   40 |                     const unsigned int width  = std::abs(rect->right - rect->left);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   41 |                     const unsigned int height = std::abs(rect->bottom - rect->top);
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   42 |
      |
   43 |                     // See http://stackoverflow.com/a/12654433/2851815 and up for DPI. In short: can't be done too too well, go with best solution.
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   44 |                     bundle.ret.push_back({width, height, monitor_dpi ? monitor_dpi : desktop_dpi, monitor_bpp ? monitor_bpp : desktop_bpp,
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   45 |                                           monitor_refresh_rate ? monitor_refresh_rate : desktop_refresh_rate});
      |                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   46 |
      |
   47 |                     return TRUE;
      |                     ~~~~~~~~~~~~
   48 |             }),
      |             ~~
make[2]: *** [CMakeFiles/infoware.dir/build.make:617: CMakeFiles/infoware.dir/src/system/displays/displays_windows.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:94: CMakeFiles/infoware.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

With:
```
$ i686-w64-mingw32-g++ --version
i686-w64-mingw32-g++ (GCC) 13.1.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

nthirtyone commented on tools/i686-w64-mingw32.cmake:
```diff
@@ -0,0 +1,14 @@
+set(TOOLCHAIN_PREFIX i686-w64-mingw32)
```
We can reduce duplication by making one of the toolchains a symlink to another and then guess `TOOLCHAIN_PREFIX` based on `CMAKE_CURRENT_LIST_FILE`.  That was my first solution, but I recalled that without Developer Mode enabled Windows users could encounter problems with the symlinks.

nabijaczleweli replied:
This and the other "toolchain" file looks, to me, like system- and build-specific config, with no business being shipped; I think move these to `.github` (`.github/workflows`?), and maybe merge them (looks like they're identical except for `TOOLCHAIN_PREFIX` and `CMAKE_SYSTEM_PROCESSOR`, which maybe can go into `-D`?) or inline them as `-D` wholesale.

nthirtyone commented on CMakeLists.txt:
```diff
@@ -47,6 +47,27 @@ if(INFOWARE_USE_OPENGL AND APPLE)
 endif()
 
+if(MINGW AND CMAKE_CROSSCOMPILING)
+	set(WINE_PREFIX "$ENV{HOME}/.wine-${TOOLCHAIN_PREFIX}" CACHE PATH "Use this directory as Wine prefix")
+	set(WINE_PATH "${CMAKE_FIND_ROOT_PATH}/bin;${CMAKE_FIND_ROOT_PATH}/lib" CACHE STRING "Use this as WINEPATH")
```
Default value for `WINE_PATH` could be adjusted to match Ubuntu. Current one is based on MSYS2, MinGW itself and e.g., Arch Linux.

nthirtyone commented on tools/wine.in:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
```
Wrapper is used because `CMAKE_CROSSCOMPILING_EMULATOR` accepts lists starting with CMake 3.15. Before, a command with arguments would be interpreted as just the name of a command.

nabijaczleweli commented:
Wait, is pci_generator being compiled for the target and then run via wine? Does CMake *really* not support compiling build-step generators for the host instead?

nthirtyone commented:
Yep, it does not. I can try switching to use `ExternalProject` hack (for the lack of better term) instead (e.g., [vtkiOS.cmake:62-77](https://gitlab.kitware.com/vtk/vtk/-/blob/master/CMake/vtkiOS.cmake#L62-77)). In short, self import as external, run whole CMake generation using different toolchain, build, get the target binaries, go back to main project.

nthirtyone commented:
It won't work with just binfmt sadly, because Wine does not use MinGW bin directories (or Ubuntu custom paths for runtime) by default and CMake does not allow propagating environmental variables from configuration stage (e.g., `ENV{WINEPATH}`) to build stage. Alternatives to the current solution with wrapping script are: prepending `${CMAKE_COMMAND} -P env ...` to custom target (and tests) or offloading all the work to system provided wrappers for toolchains (e.g., i686-w64-mingw-{cmake,wine} commands; Ubuntu does not have such).

nthirtyone commented:
Rebase might be done to hastily, I don't know what's your preferred workflow, sorry. I removed wine dependency altogether and switched to use ExternalProject as discussed. The conditions are made to not be MinGW dependent, so naively it should work in other cross-compilation environments, too. If we remove MinGW CI jobs, I could remove the toolchains, too.

nabijaczleweli commented on CMakeLists.txt:
```diff
@@ -147,7 +157,11 @@ if(WIN32 AND NOT MSVC)
+	if(NOT MINGW)
+		target_link_libraries(infoware PRIVATE gdi32 version Ole32 OleAut32 wbemuuid ntdll)
+	else()
+		target_link_libraries(infoware PRIVATE gdi32 version ole32 oleaut32 wbemuuid ntdll)
```
Would this not also work without the conditional by always requesting all-lowercase?

nthirtyone replied:
That appears to be the case.

nabijaczleweli commented:
Looks fine, CI passes, works for me on native builds; thanks!

nthirtyone commented:
Cheers!
